File: Algorithms/20. Valid Parentheses/python/20_valid_parentheses.py
from cgi import test


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        
        for char in s:
            if char=='(' or char=='[' or char=='{':
                stack.append(char)
            
            try:
                if char==')':
                    if stack.pop() != '(':
                        return False
                elif char==']':
                    if stack.pop() != '[':
                        return False
                elif char=='}':
                    if stack.pop() != '{':
                        return False
            except:
                return False
        
        if len(stack) != 0:
            return False
        return True
        
    
sol = Solution()

test1 = '()'
test2 = '()[]{}'
test3 = '(]'
test4 = '['

print(sol.isValid(test1))
print(sol.isValid(test2))
print(sol.isValid(test3))
print(sol.isValid(test4))

File: Algorithms/20. Valid Parentheses/java/20_valid_parentheses.java
import java.util.ArrayList;
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        // Stack< x = new ArrayList<>()
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(')');
            }else if (s.charAt(i) == '['){
                stack.push(']');
            }else if (s.charAt(i) == '{'){
                stack.push('}');
            } else if (stack.isEmpty() || stack.pop() != s.charAt(i)){
                return false;
            }
        }
        // System.out.println(stack);
        return stack.isEmpty();
    }

public static void main(String[] args) {
    Solution sol = new Solution();

    String test1 = new String("()");
    String test2 = new String("()[]{}");
    String test3 = new String("(]");
    String test4 = new String("[");
    String test5 = new String("]");

    System.out.println(sol.isValid(test1));
    System.out.println(sol.isValid(test2));
    System.out.println(sol.isValid(test3));
    System.out.println(sol.isValid(test4));
    System.out.println(sol.isValid(test5));
    }
}